src/main/java/com/smartbear/readyapi/client/teststeps/request/HttpRequestStepBuilder.java

```diff
@@ -1,87 +1,67 @@
+import com.smartbear.readyapi.client.model.RequestTestStepBase;
 import com.smartbear.readyapi.client.teststeps.TestStepBuilder;
-import com.smartbear.readyapi.client.teststeps.TestStepTypes;
-import com.smartbear.readyapi.client.teststeps.TestSteps;
 
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
-import static com.smartbear.readyapi.client.Validator.validateNotEmpty;
 import static com.smartbear.readyapi.client.assertions.Assertions.contains;
 import static com.smartbear.readyapi.client.assertions.Assertions.invalidStatusCodes;
-import static com.smartbear.readyapi.client.assertions.Assertions.jsonPathContent;
-import static com.smartbear.readyapi.client.assertions.Assertions.jsonPathCount;
 import static com.smartbear.readyapi.client.assertions.Assertions.notContains;
 import static com.smartbear.readyapi.client.assertions.Assertions.responseSLA;
 import static com.smartbear.readyapi.client.assertions.Assertions.script;
 import static com.smartbear.readyapi.client.assertions.Assertions.validStatusCodes;
 import static com.smartbear.readyapi.client.assertions.Assertions.xPathContains;
 import static com.smartbear.readyapi.client.assertions.Assertions.xQueryContains;
 
-public class BaseRestRequest<RequestBuilderType extends RestRequestBuilder> implements TestStepBuilder<RestTestRequestStep>, RestRequestBuilder<RequestBuilderType> {
-    protected RestTestRequestStep testStep = new RestTestRequestStep();
+abstract public class HttpRequestStepBuilder<RequestBuilderType extends HttpRequestStepBuilder, RequestTestStepType extends RequestTestStepBase> implements TestStepBuilder {
+
+    private final RequestTestStepType testStep;
     private List<Parameter> parameters = new ArrayList<>();
     private List<AssertionBuilder> assertionBuilders = new ArrayList<>();
     private Map<String, Object> headers = new HashMap<>();
 
-    public enum ParameterType {
-        MATRIX, HEADER, QUERY, PATH
-    }
-
-    public BaseRestRequest(String uri, TestSteps.HttpMethod method) {
-        testStep.setURI(uri);
-        testStep.setMethod(method.toString());
-    }
-
-    @Override
-    public RequestBuilderType named(String name) {
-        testStep.setName(name);
-        return (RequestBuilderType) this;
+    protected HttpRequestStepBuilder(RequestTestStepType testStep, String type) {
+        this.testStep = testStep;
+        this.testStep.setType(type);
     }
 
-    @Override
-    public RequestBuilderType addQueryParameter(String parameterName, String value) {
-        return addParameter(parameterName, value, ParameterType.QUERY);
+    final protected RequestTestStepType getTestStep() {
+        return testStep;
     }
 
-    @Override
-    public RequestBuilderType addPathParameter(String parameterName, String value) {
-        return addParameter(parameterName, value, ParameterType.PATH);
+    protected List<Parameter> getParameters() {
```

Comment:
I think we're overdoing the inheritance here. Maybe this should be pushed down into the subclasses, since REST and SOAP parameters are very different things.